Checksum calculating storage wrapper: Initial draft #26655

IljaN opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Commit:
External storage tables are now created in a migration

IljaN committed:
Sharing dialog: Names sorting is case sensitive #25971

DeepDiver1975 committed:
The default collation is set in the connection for schema operations

VicDeo committed:
Emit postSetEnabled. Closes #23970

VicDeo committed:
Dispatch postSetEnabled

Commit:
Integration test check download without saving file locally

Use Guzzle stream mode to download the contents instead of using a
temporary local file.

Commit:
Properly handle deleted users in UserMountCache

When encountering entries from a deleted user, discard such entries.

IljaN committed:
Sharing dialog: Names sorting is case sensitive: adding tests #25971

DeepDiver1975 committed:
Kill doctrine/migrations

DeepDiver1975 committed:
Adding schema methods to IDbConnection

DeepDiver1975 committed:
Introduce own migration mechanism

Implement occ commands migrations:status, migrations:execute, migrations:generate and migrations:migrate

DeepDiver1975 committed:
Implement Version20170111103310 as ISchemaMigration

DeepDiver1975 committed:
Fix OracleMigrator to set proper quoted names

IljaN committed:
Checksum calculating storage wrapper: Initial draft #26655

IljaN committed:
Checksum calculating storage wrapper: Initial draft #26655

IljaN commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,82 @@
+	public function rename($path1, $path2) {
+		if (array_key_exists($path1, $this->fileHashes)) {
+			$checksum = $this->fileHashes[$path1];
+			// Update checksum in oc_filecache here?
```
Inject Storage Layer and update oc_filechace here?

PVince81 replied:
Hmmm, no. When renaming a non-part file, this should already happen automatically thorugh the `Cache\Updater` class.
You can confirm this by manually storing a value in the checksum column with a DB client, then renaming the entry.

The question we had was about part files, so might be more tricky / another location.

PVince81 commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,82 @@
+	 * @return bool
+	 */
+	private static function requiresChecksum($path, $mode) {
+		return substr($path, 0, 6) == 'files/' && $mode != 'r';
```
always use strict equality `===` and `!==`

PVince81 commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,82 @@
+namespace OC\Files\Storage\Wrapper;
+
+
+class Checksum extends Wrapper {
```
PHPDoc

PVince81 commented on lib/private/Files/Stream/Filter/Sha1.php:
```diff
@@ -0,0 +1,71 @@
+	function onClose () {
+		$hash = hash_final($this->ctx);
+		// Debug
+		file_put_contents(
```
we usually use logging for debugging, something like:
```
	\OCP\Util::writeLog('DEBUG', 'some text', \OCP\Util::DEBUG);
```

Commit:
[tx-robot] updated from transifex

PVince81 reviewed (commented):
Need to think about how to store the checksum in the file cache.

A quick way would be to use the matching `File\Cache` instance and store it there. But that would cause an additional insert. Note that the encryption wrapper already does something similar with the size, so maybe it's ok to do so...

DeepDiver1975 committed:
Kill that . (#27046)

VicDeo committed:
Do not use listener in test

Commit:
[tx-robot] updated from transifex

Commit:
[tx-robot] updated from transifex

Commit:
Merge pull request #27028 from owncloud/usermountcache-deletedusers

Properly handle deleted users in UserMountCache

IljaN committed:
Checksum calculating storage wrapper: Initial draft - rewrite to stream wrapper #26655

IljaN committed:
Checksum calculating storage wrapper: Initial draft - rewrite to stream wrapper #26655

VicDeo committed:
Pass EventDispatcher

Commit:
[tx-robot] updated from transifex

Commit:
Add integration test for trashbin

Add test for basic deletion.
Add test when deleting from shared folder as recipient.
Add test to check that metadata stays when moving out of shared folder
as recipient.

DeepDiver1975 committed:
Merge pull request #27041 from owncloud/owncloud-migrations

ownCloud migrations

IljaN committed:
Upload and download checksum is working #26655

IljaN committed:
Upload and download checksum is working #26655

Commit:
Merge pull request #26950 from IljaN/25971_sharing_diag_case_insensetive

Sharing dialog: Names sorting is case sensitive #25971

Commit:
Merge pull request #27069 from owncloud/inttest-trashbin-share

Add integration test for trashbin

IljaN committed:
Fixing occhecksum already defined error #26655

IljaN committed:
Fixing occhecksum already defined error #26655

Commit:
Merge pull request #26984 from owncloud/dispatch-endisable

Emit postSetEnabled for user

Commit:
Merge pull request #27008 from owncloud/inttest-checkurlwithoutsaveto

Integration test check download without saving file locally

PVince81 commented on lib/private/Files/Cache/Scanner.php:
```diff
@@ -202,7 +202,7 @@ public function scanFile($file, $reuseExisting = 0, $parentId = -1, $cacheData =
 				if (!empty($newData)) {
 					// Reset the checksum if the data has changed
-					$newData['checksum'] = '';
+					unset($newData['checksum']);
```
I think  `addToCache` only writes the values that are inside the array. So unsetting the checksum will not overwrite an existing "checksum" value if one exists. So you might want to revert this change.

However I see that this might be the block for "new data to insert" in which case not having a checksum in the data would also not set the checksum to anything.

Can you confirm if my reasoning is correct ?

IljaN replied:
@PVince81 I see what you mean... I will check if there are any bad effects trough this change

This change is required to make the  [partial data merge](https://github.com/IljaN/core/blob/9c1147b69f17fe3c4ff0773f94e4057bf40bd2b1/lib/private/Files/Cache/Cache.php#L238) work. array_merge(['checksum' => 'deadbeef', 'checksum' => '] equals ['checksum' => '' "] :(, the checksum gets overwritten by the empty checksum. :(

PVince81 commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,100 @@
+
+use OC\Files\Stream\Checksum as ChecksumStream;
+
+class Checksum extends Wrapper {
```
PHPDoc, what does this class do ?

IljaN replied:
Will do...

PVince81 commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,100 @@
+		$renameOk = $this->getWrapperStorage()->rename($path1, $path2);
+		$meta = $this->getMetaData($path1);
+
+		if (isset($meta['checksum'])) {
```
I'm a bit surprised that this is needed. I thought there was another location in the code that would anyway move over the metadata when renaming.

Did you test and saw that it did not work without your addition ? Something to investigate later.

Note that calling `getCache()->put()` here would produce an additional SQL query just for the checksum. So if we find that there is a way to add the checksum in the props from the regular rename, it might save one SQL query.

IljaN replied:
@PVince81 
I just have the stream of the partfile in the put (on which I need to calculate the checksum), I can't write the checksum of the partfile in to oc_filecache because there are no part-files in oc_filecache right? So I need to save the hash of the partfile and map it to the original filename somehow.

 There is no additional query: The updater collects all the updates and then executes one query after the all data is complete. :open_mouth: 

See:https://github.com/IljaN/core/blob/9c1147b69f17fe3c4ff0773f94e4057bf40bd2b1/lib/private/Files/Cache/Cache.php#L237-L244

PVince81 replied:
Now that you say it, I remember that behavior about the cache. It can have "incomplete" entries and only saves at the end. Quite quirky and unintuitive. Anyway, that will do for now.

PVince81 commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,100 @@
+	 * @return bool
+	 */
+	public function file_put_contents($path, $data) {
+		return parent::file_put_contents($path, $data); // TODO: Change the autogenerated stub
```
likely replace with a wrapped `fopen` call that saves the whole data with a `streamCopy` ?

PVince81 commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,100 @@
+	 */
+	public function getMetaData($path) {
+		$parentMetaData = parent::getMetaData($path);
+		$checksums = ChecksumStream::getChecksums();
```
I'd suggest `getChecksums($path)` directly. Make that function return null or empty string if there was no checksum.

This way this class doesn't need to know about the internals of `ChecksumStream`'s static stuff.

PVince81 commented on lib/private/Files/Stream/Checksum.php:
```diff
@@ -0,0 +1,130 @@
+	private static $checksums = [];
+
+	public function __construct() {
+		$this->hashCtx = hash_init('sha1');
```
Maybe make this configurable through constructor args ? (optional)

radhefa replied:
Radhefa

PVince81 commented on lib/private/Files/Stream/Checksum.php:
```diff
@@ -0,0 +1,130 @@
+	 */
+	public function stream_open($path, $mode, $options, &$opened_path) {
+		$context = parent::loadContext('occhecksum');
+        $this->setSourceStream($context['source']);
```
bad indent

PVince81 commented on lib/private/Files/Stream/Checksum.php:
```diff
@@ -0,0 +1,130 @@
+	 * @return bool
+	 */
+	public function stream_close() {
+        self::$checksums[$this->getPathFromContext()] = hash_final($this->hashCtx);
```
another bad indent

PVince81 commented on lib/private/Files/Stream/Checksum.php:
```diff
@@ -0,0 +1,130 @@
+	/**
+	 * @return array
+	 */
+	public static function getChecksums() {
```
`$path` param

PVince81 reviewed (commented):
I haven't tested this yet. The code looks good so far with a few minor comments.

IljaN commented on lib/private/Files/Storage/Wrapper/Checksum.php:
```diff
@@ -0,0 +1,100 @@
+	 * @return bool
+	 */
+	private static function requiresChecksum($path) {
+		return substr($path, 0, 6) === 'files/';
```
@PVince81 Can 'files/' stay hardcoded?

PVince81 replied:
It's ugly but I don't see any other way.

Unless you want to introduce a new class with constants ? The encryption code and others also use this hard-coded piece.

IljaN committed:
Merge remote-tracking branch 'origin/26655_ComputeChecksumWrapper' into 26655_ComputeChecksumWrapper

IljaN committed:
Implemented comparison for normal and chunked, simplified implementation, parametrized hash algo #26655

IljaN commented:
Moved to https://github.com/owncloud/core/pull/27097 due to broken branch
